### PR TITLE
HashSet implementation

### PR DESCRIPTION
**Pull request by ruslansennov**

### CODE REVIEW

File: src/main/java/javaslang/collection/HashArrayMappedTrie.java [hunk 1]
```diff
@@ -17,7 +17,7 @@
 /**
  * <a href="https://en.wikipedia.org/wiki/Hash_array_mapped_trie">Hash array mapped trie (HAMT)</a>.
  */
-abstract class HashArrayMappedTrie<K, V> implements Iterable<Tuple2<K, V>> {
+abstract class HashArrayMappedTrie<K, V> implements Iterable<Tuple2<K, V>>, Serializable {
```
Comment (danieldietrich) [inline src/main/java/javaslang/collection/HashArrayMappedTrie.java]: Hi, I think the build fails because the serialVersionUID = 1L is missing...

**Comment by danieldietrich** [comment]
Wow, thanks! That was fast :)

File: src/main/java/javaslang/collection/HashArrayMappedTrie.java (hunk 1 above)
Comment (ruslansennov) [inline src/main/java/javaslang/collection/HashArrayMappedTrie.java]: Yes, I see, fixed.
But there is another error message when I do <code>mvn test</code>:
```text
Failed tests:   shouldHaveAConsistentTypeSystem(javaslang.TypeConsistencyTest): Unoverriden methods found. Check if this is ok and copy & paste the necessary lines into the TypeConsistencyTest.WHITELIST:
```

File: src/main/java/javaslang/collection/HashSet.java [hunk 2]
```diff
@@ -0,0 +1,515 @@
...
+            return tree.isEmpty() ? empty() : new HashSet<>(tree);
+        }
+    }
+
+    private final static Object O = new Object();
+
+    private final HashArrayMappedTrie<T, Object> tree;
+    private final Lazy<List<T>> list;
+    private final Lazy<Integer> hash;
+
```
Comment (danieldietrich) [inline src/main/java/javaslang/collection/HashSet.java]: With `list` we trade space against performance here. This is nice regarding performance. I wonder how Scala does this for Map/Set... I'm just curious.

File: src/main/java/javaslang/collection/HashArrayMappedTrie.java (hunk 1 above)
Comment (danieldietrich) [inline src/main/java/javaslang/collection/HashArrayMappedTrie.java]: Ah - there is a unit test that checks, if all return types are as special as they can be.

Example:

Option.map returns Option but we know that None.map returns a None.

If we do not overwrite Option.map within None that way, it is put on the WHITELIST in the unit test.

You can copy the String 1:1 from the console and add a WHITELIST entry.

File: src/main/java/javaslang/collection/HashSet.java (hunk 2 above)
Comment (danieldietrich) [inline src/main/java/javaslang/collection/HashSet.java]: But this prototype is fine. It is exactly what we wanted!

File: src/main/java/javaslang/collection/HashArrayMappedTrie.java (hunk 1 above)
Comment (ruslansennov) [inline src/main/java/javaslang/collection/HashArrayMappedTrie.java]: I prefer to change the signature :)

**Pull request merged**